ADR/0032-decoupling-deployment.md

```diff
@@ -0,0 +1,211 @@
+- Decoupling [integration-service] from the [application-service] APIs, like [Application] and
+  [Component]. It will still promote images to the "global candidate list" (on the [Component]
+  resources) and continue to use the list of [Components] to guide its construction of [Snapshots].
+
+### Use Case Descriptions
+
+**During onboarding**: whereas today when a user requests a new appstudio tier namespace, the tier
+template includes an [Environment] that the integration-service will promote to. Tomorrow, the
+appstudio tier template should no longer include an [Environment] which on its own will cause
+integration-service to _not_ trigger a deployment when testing completes. Instead, the appstudio
+tier template should include a [ReleasePlan] with a reference to the [push-to-registry]
+release pipeline. This new default [ReleasePlan] should carry parameters such that whenever
+a [Snapshot] is successfully tested, a [Release] is created that re-tags the images in build-time
+quay repositories with a tag like `:released` or `:validated` (name tbd). The
+[push-to-registry] pipeline can use the `appstudio-pipeline` service account in the user's
+dev workspace, which already has push access to the repository in question.
+
+```mermaid
+flowchart TD
+    NSTierTemplate --> |provides| ReleasePlan
+    A[fa:fa-code-commit Git Commit] --> |webhook| PipelinesAsCode
+    PipelinesAsCode --> |creates| build[Build PipelineRun]
+    build --> |triggers| integration-service
+    integration-service --> |asks| decision{Is there a ReleasePlan?}
+    ReleasePlan --> |influences| decision
+    decision --> |yes| Release
+    build --> |pushes the image| quay.io/redhat-user-workloads/tenant/component
+    Release --> |tags the image for release| quay.io/redhat-user-workloads/tenant/component
+    quay.io/redhat-user-workloads/tenant/component --> renovatebot[User provided RenovateBot]
+    renovatebot --> |updates| gitops-repo[User provided GitOps repo]
+    gitops-repo --> |causes deployment via| argocd[User provided ArgoCD]
+    argocd --> |deploys to| cluster[User provided cluster]
+```
+
+Outside of the AppStudio member cluster, the user is responsible for acquiring a gitops repo and
+deployment environments of their choice, manually laying out their application resources in the repo
+(assisted by tools like `kam`), specifying image references by tag to match the `:released` or
+`:validated` tagging schem mentioned above, configuring Argo to deploy from their gitops repo, and
+configuring [renovatebot] to propagate image updates by digest to their gitops repo.
+
+**For manual creation of new environments** - the user manages this directly using a combination of
+their gitops repo and argo, outside of the AppStudio member cluster.
+
+**For automated testing in ephemeral environments** - the user specifies an
+[IntegrationTestScenario] CR with a [DeploymentTargetClass] reference. After a build completes, but
```

Comment:
The user should specify a DeploymentTargetClaim (the class is mentioned in the claim). The integration service should create the DTC as mentioned in the IntegrationTest and deletes it when the test pipeline completes.

Reply:
I think you're right. They need to specify a DeploymentTargetClaim *spec*. The integration-service should create a real DTC from that spec.

Reply:
Something like:

```mermaid
flowchart TD
    User --> |provides| IntegrationTestScenario
    IntegrationTestScenario --> |embeds| spec[DeploymentTargetClaim spec]
    commit[fa:fa-code-commit Git Commit] --> |webhook| PipelinesAsCode
    PipelinesAsCode --> |creates| build[Build PipelineRun]
    build --> |triggers| integration-service
    integration-service --> |consults| IntegrationTestScenario
    spec --> |is used to create| DeploymentTargetClaim
    integration-service --> |creates| DeploymentTargetClaim
    DeploymentTargetClaim --> |is bound to| DeploymentTarget
    DeploymentTarget --> |is associated with| kubeconfig[kubeconfig Secret]
    kubeconfig --> |is provided to| test
    integration-service --> |creates| test[Test PipelineRun]
    DeploymentTarget --> |represents| cluster
    test --> |deploys user app to| cluster
    test --> |executes tests against| cluster
```

Reply:
Updated in 382bcb9.